### PR TITLE
CSA-62/searchbar

### PR DESCRIPTION
**Pull request by tymnik**
- component SearchBar;
- SearchBar redux (roadmaps/search);
- index.ts added where it was absent for reexports;
- icons access/deny added to sprite;

### CODE REVIEW

File: src/components/Header/SearchBar/SearchBar.module.css [hunk 1]
```diff
@@ -53,4 +59,25 @@
   height: 28px;
   fill: none;
   stroke: var(--text-primary);
+  position: relative;
+}
+
+.dropdown {
+  position: absolute;
+  border-bottom: 1px solid #f4f4f4;
```
Comment (VovaLyapota) [inline src/components/Header/SearchBar/SearchBar.module.css]: it`s needed to add color to vars

File: src/components/Header/SearchBar/SearchBar.tsx [hunk 2]
```diff
@@ -1,46 +1,112 @@
...
+  className: string;
 }
 
-const SearchBar: React.FC<SearchBarProps> = ({ onClose }) => {
-  const [isSearchOpen, setIsSearchOpen] = useState(true);
+const SearchBar = ({ onClose, className }: SearchBarProps) => {
+  const dispatch = useAppDispatch();
+  const roadmaps = useAppSelector(selectSearchList);
+  const [searchTerm, setSearchTerm] = useState('');
+  const [selectedSuggestion, setSelectedSuggestion] = useState<string>('');
```
Comment (VovaLyapota) [inline src/components/Header/SearchBar/SearchBar.tsx]: there is no need to add generic in useState, it automatically sets during declaration

File: src/components/Header/SearchBar/SearchBar.tsx [hunk 3]
```diff
@@ -1,46 +1,112 @@
...
+      ) {
+        onClose();
+      }
+    };
+
+    document.addEventListener('mousedown', handleClickOutside);
+    return () => {
+      document.removeEventListener('mousedown', handleClickOutside);
+    };
+  }, [onClose]);
```
Comment (VovaLyapota) [inline src/components/Header/SearchBar/SearchBar.tsx]: I think you don't need to check onClose fn in useEffect

File: src/store/roadmaps/slice.ts [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+      .addCase(searchRoadmaps.fulfilled, onSuccess)
+      .addCase(searchRoadmaps.rejected, onRejected);
+  },
+});
+
+const onSuccess = (
+  state: RoadmapState,
+  action: PayloadAction<SearchListResponse>,
+) => {
+  console.log('Received payload:', action.payload);
```
Comment (VovaLyapota) [inline src/store/roadmaps/slice.ts]: We don't need a console

File: src/store/roadmaps/slice.ts [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+};
+
+const onPending = (state: RoadmapState) => {
+  state.isLoading = true;
+  state.error = null;
+};
+
+const onRejected = (
+  state: RoadmapState,
+  action: PayloadAction<string | undefined>,
```
Comment (VovaLyapota) [inline src/store/roadmaps/slice.ts]: If you have a rejected action, you'll have had a string in rejectWithValue func, and you shouldn't write  <string | undefined>, use ! operator on 54 line instead (action.payload!) - it guaranties that payload is always in reject action

File: src/components/Header/Header.tsx [hunk 6]
```diff
@@ -35,28 +35,17 @@ const Header = () => {
...
-    return () => {
-      document.removeEventListener('keydown', handleKeyDown);
-    };
-  }, []);
-
   return (
     <header className={styles.header}>
-      <a href={'/'} target="_blank" rel="noopener noreferrer">
+      <a href={'/'} rel="noopener noreferrer">
         <img src={logo} alt="Career Skill Atlas logo" />
```
Comment (StanislavKharkiv) [inline src/components/Header/Header.tsx]: Why don't use ROUTES.HOME ?

File: src/components/Header/SearchBar/SearchBar.tsx [hunk 7]
```diff
@@ -1,46 +1,113 @@
...
-  const handleSubmit = (event: React.FormEvent<HTMLFormElement>) => {
-    event.preventDefault();
+  const filteredRoadmaps = roadmaps.filter((roadmap) =>
+    roadmap.title.toLowerCase().includes(searchTerm.toLowerCase()),
+  );
+
+  const throttledSearch = useCallback(
+    throttle(1000, (value: string) => dispatch(searchRoadmaps(value))),
+    [dispatch],
+  );
```
Comment (StanislavKharkiv) [inline src/components/Header/SearchBar/SearchBar.tsx]: why you need useCallback here? (and where array with dependencies?)

File: src/components/Header/SearchBar/SearchBar.tsx [hunk 8]
```diff
@@ -1,46 +1,113 @@
...
-    setIsSearchOpen(!isSearchOpen);
+    setSelectedSuggestion('');
   };
 
+  useEffect(() => {
+    const handleClickOutside = (event: MouseEvent) => {
+      if (
+        searchBarRef.current &&
+        !searchBarRef.current.contains(event.target as Node)
+      ) {
```
Comment (StanislavKharkiv) [inline src/components/Header/SearchBar/SearchBar.tsx]: just use optional chaining `!searchBarRef.current?.contains(event.target as Node)`

File: src/api/roadmap.ts [hunk 9]
```diff
@@ -1,10 +1,15 @@
...
-  return data;
+  try {
+    const response = await axios.get(
+      `${ENDPOINTS.ROADMAPS}search/?search=${searchText}`,
+    );
+    return response;
+  } catch (error) {
+    throw (error as AxiosError).message || 'Search failed: Unknown error';
+  }
 };
```
Comment (StanislavKharkiv) [inline src/api/roadmap.ts]: We don't need try catch here. And forget fix path `search?search=${searchText}`

**Review by StanislavKharkiv — COMMENTED** [review]
You can change on BE to CSA-64 branch and test your requests. Ask BE how add some for response and why they don't give answer when user not logged. I think search should work for all users even for guests. Check this please

**Review by StanislavKharkiv — APPROVED** [review]

**Pull request merged**